File: 3.x.x/crates-utopia/queue/src/message.rs
use serde_json::{json, Map, Value};

/// Envelope stored on a broker (`pid`, `queue`, `timestamp`, `payload`, `attempts`).
///
/// PHP `Utopia\Queue\Message`.
#[derive(Debug, Clone, PartialEq)]
pub struct Message {
    pid: String,
    queue: String,
    timestamp: i64,
    payload: Option<Value>,
    attempts: i64,
}

impl Default for Message {
    fn default() -> Self {
        Self::new()
    }
}

impl Message {
    pub fn new() -> Self {
        Self {
            pid: String::new(),
            queue: String::new(),
            timestamp: 0,
            payload: None,
            attempts: 0,
        }
    }

    /// Build from a JSON object (PHP constructor `array $array`).
    pub fn from_value(value: &Value) -> Self {
        let Some(obj) = value.as_object() else {
            return Self::new();
        };
        if obj.is_empty() {
            return Self::new();
        }
        Self {
            pid: obj
                .get("pid")
                .and_then(Value::as_str)
                .unwrap_or_default()
                .to_owned(),
            queue: obj
                .get("queue")
                .and_then(Value::as_str)
                .unwrap_or_default()
                .to_owned(),
            timestamp: json_i64(obj.get("timestamp")).unwrap_or(0),
            payload: obj.get("payload").cloned().or_else(|| Some(json!({}))),
            attempts: json_i64(obj.get("attempts")).unwrap_or(0),
        }
    }

    pub fn set_pid(&mut self, pid: impl Into<String>) -> &mut Self {
        self.pid = pid.into();
        self
    }

    pub fn set_queue(&mut self, queue: impl Into<String>) -> &mut Self {
        self.queue = queue.into();
        self
    }

    pub fn set_timestamp(&mut self, timestamp: i64) -> &mut Self {
        self.timestamp = timestamp;
        self
    }

    pub fn set_payload(&mut self, payload: Value) -> &mut Self {
        self.payload = Some(payload);
        self
    }

    pub fn get_pid(&self) -> &str {
        &self.pid
    }

    pub fn get_queue(&self) -> &str {
        &self.queue
    }

    pub fn get_timestamp(&self) -> i64 {
        self.timestamp
    }

    pub fn get_payload(&self) -> Value {
        self.payload.clone().unwrap_or(Value::Null)
    }

    pub fn get_payload_ref(&self) -> Option<&Value> {
        self.payload.as_ref()
    }

    pub fn get_attempts(&self) -> i64 {
        self.attempts
    }

    pub fn set_attempts(&mut self, attempts: i64) -> &mut Self {
        self.attempts = attempts;
        self
    }

    /// PHP `asArray()` keys: `pid` / `queue` / `timestamp` / `payload` / `attempts`.
    pub fn as_array(&self) -> Value {
        json!({
            "pid": self.pid,
            "queue": self.queue,
            "timestamp": self.timestamp,
            "payload": self.payload.clone().unwrap_or(Value::Null),
            "attempts": self.attempts,
        })
    }

    pub fn as_object(&self) -> Map<String, Value> {
        self.as_array().as_object().cloned().unwrap_or_default()
    }
}

fn json_i64(value: Option<&Value>) -> Option<i64> {
    match value? {
        Value::Number(n) => n.as_i64().or_else(|| n.as_f64().map(|f| f as i64)),
        Value::String(s) => s.parse().ok(),
        _ => None,
    }
}

File: src/Queue/Message.php
<?php

namespace Utopia\Queue;

class Message
{
    protected string $pid;
    protected string $queue;
    protected int $timestamp;
    protected array $payload;
    protected int $attempts = 0;
    protected ?int $sequence = null;
    protected bool $terminal = false;

    public function __construct(array $array = [])
    {
        if ($array === []) {
            return;
        }

        $this->pid = $array['pid'];
        $this->queue = $array['queue'];
        $this->timestamp = $array['timestamp'];
        $this->payload = $array['payload'] ?? [];
        $this->attempts = $array['attempts'] ?? 0;
        $this->sequence = $array['sequence'] ?? null;
    }

    private ?string $receipt = null;

    public function getReceipt(): ?string
    {
        return $this->receipt;
    }

    public function setReceipt(string $receipt): self
    {
        $this->receipt = $receipt;
        return $this;
    }

    public function setPid(string $pid): self
    {
        $this->pid = $pid;

        return $this;
    }

    public function setQueue(string $queue): self
    {
        $this->queue = $queue;

        return $this;
    }

    public function setTimestamp(int $timestamp): self
    {
        $this->timestamp = $timestamp;

        return $this;
    }

    public function setPayload(array $payload): self
    {
        $this->payload = $payload;

        return $this;
    }

    public function getPid(): string
    {
        return $this->pid;
    }

    public function getQueue(): string
    {
        return $this->queue;
    }

    public function getTimestamp(): int
    {
        return $this->timestamp;
    }

    public function getPayload(): array
    {
        return $this->payload;
    }

    /**
     * Times this message has been requeued after a failed or stranded run.
     */
    public function getAttempts(): int
    {
        return $this->attempts;
    }

    public function setAttempts(int $attempts): self
    {
        $this->attempts = $attempts;

        return $this;
    }

    /**
     * The broker's own position for this message, where it has one.
     *
     * A stronger deduplication key than the pid for a handler that needs one.
     * The pid identifies the logical message and is stable across every
     * redelivery, which is what makes it the right key for "have I already done
     * this work"; the sequence identifies the stored copy, so it distinguishes
     * one delivery of a message from another. Null on brokers with no such
     * notion, so a handler must treat it as optional.
     */
    public function getSequence(): ?int
    {
        return $this->sequence;
    }

    public function setSequence(?int $sequence): self
    {
        $this->sequence = $sequence;

        return $this;
    }

    /**
     * Declare that this message must not be delivered again.
     *
     * The verdict a handler reaches about its own failure: this payload will
     * fail the same way on every attempt, so spending the redelivery budget on
     * it buys nothing and — on a broker that counts an unacked message against a
     * ceiling, as JetStream does — costs a delivery slot for the whole of it.
     * What it costs to ignore decides what a broker does with it.
     * {@see Broker\Nats::reject()} dead-letters a terminal message at once
     * instead of scheduling the next attempt, because every further attempt
     * would hold an ack slot for the length of its backoff.
     * {@see Broker\Redis::reject()} has no such ceiling to protect and leaves
     * the message on the failed list with every other rejection: nothing re-runs
     * it there either, and that list is the only one an operator can recover
     * from. A broker with no notion of the distinction rejects it the way it
     * always has.
     *
     * Throwing {@see PermanentFailure} sets this and is the shorter route. This
     * is here for a handler that cannot: the exception type belongs to a library,
     * or the classification happens somewhere that only has the message. It is
     * read when the message is rejected, so set it before the handler gives up.
     *
     * Per delivery, not part of the envelope: a redelivered copy of the same
     * payload arrives with a fresh verdict, because the reason it failed may
     * have been fixed in between.
     */
    public function terminal(bool $terminal = true): self
    {
        $this->terminal = $terminal;

        return $this;
    }

    /** Whether the handler declared this failure permanent. {@see self::terminal()} */
    public function isTerminal(): bool
    {
        return $this->terminal;
    }

    public function asArray(): array
    {
        return [
            'pid' => $this->pid,
            'queue' => $this->queue,
            'timestamp' => $this->timestamp,
            'payload' => $this->payload ?? null,
            'attempts' => $this->attempts,
            'sequence' => $this->sequence,
        ];
    }
}
